src/erlcloud_waf.erl

```diff
@@ -0,0 +1,1191 @@
+        list_web_acls/0, list_web_acls/1, list_web_acls/2, list_web_acls/3,
+        update_byte_match_set/3, update_byte_match_set/4,
+        update_ip_set/3, update_ip_set/4,
+        update_rule/3, update_rule/4,
+        update_size_constraint_set/3, update_size_constraint_set/4,
+        update_sql_injection_match_set/3, update_sql_injection_match_set/4,
+        update_web_acl_default_action/3, update_web_acl_default_action/4,
+        update_web_acl/3, update_web_acl/4
+        ]).
+
+
+-define(API_VERSION, "20150824").
+-define(LIMIT_MAX, 100).
+
+%%%------------------------------------------------------------------------------
+%%% Shared types
+%%%------------------------------------------------------------------------------
+%% -type waf_opts() :: list().
+
+-type waf_return_val() :: {ok, proplists:proplist()} | {error, term()}.
+
+-export_type([waf_return_val/0]).
+
+%%%------------------------------------------------------------------------------
+%%% Library initialization.
+%%%------------------------------------------------------------------------------
+
+-spec new(string(), string()) -> aws_config().
+
+new(AccessKeyID, SecretAccessKey) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey
+      }.
+
+-spec new(string(), string(), string()) -> aws_config().
+
+new(AccessKeyID, SecretAccessKey, Host) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey,
+       waf_host=Host
+      }.
+
+-spec new(string(), string(), string(), non_neg_integer()) -> aws_config().
+
+new(AccessKeyID, SecretAccessKey, Host, Port) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey,
+       waf_host=Host,
+       waf_port=Port
+      }.
+
+-spec configure(string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey)),
+    ok.
+
+-spec configure(string(), string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host, Port) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port)),
+    ok.
+
+default_config() ->
+    erlcloud_aws:default_config().
+
+
+-spec waf_result_fun/1 ::
+          (Request :: aws_request()) -> aws_request().
+waf_result_fun(#aws_request{response_type = ok} = Request) ->
+    Request;
+waf_result_fun(#aws_request{response_type = error,
+                                  error_type = aws,
+                                  response_status = Status} = Request) when Status >= 500 ->
+    Request#aws_request{should_retry = true};
+waf_result_fun(#aws_request{response_type = error, error_type = aws} = Request) ->
+    Request#aws_request{should_retry = false}.
```

Comment:
It'll be better to place this function after `waf_request_no_update/3` where is its first usage.